include/aikido/util/detail/yaml_extension.hpp

```diff
@@ -0,0 +1,246 @@
+  }
+  else if (node.Tag() == "Matrix" || node.Tag() == "!Matrix")
+  {
+    const auto cols = node[0].size();
+
+    if (MatrixType::ColsAtCompileTime != Eigen::Dynamic
+        && cols != MatrixType::ColsAtCompileTime)
+    {
+      std::stringstream ss;
+      ss << "Matrix has incorrect number of cols: expected "
+         << MatrixType::ColsAtCompileTime << "; got " << cols << ".";
+      throw runtime_error(ss.str());
+    }
+
+    matrix.resize(rows, cols);
+
+    for (auto r = 0u; r < node.size(); ++r)
+    {
+      if (node[r].Type() != YAML::NodeType::Sequence)
+      {
+        std::stringstream ss;
+        ss << "Row " << r << " of the matrix must be a sequence.";
+        throw runtime_error(ss.str());
+      }
+      else if (node[r].size() != cols)
+      {
+        std::stringstream ss;
+        ss << "Expected row " << r << " to have " << cols << " columns; got "
+           << node[r].size() << ".";
+        throw runtime_error(ss.str());
+      }
+
+      for (auto c = 0u; c < cols; ++c)
+        matrix(r, c) = node[r][c].template as<_Scalar>();
+    }
+  }
+  else
+  {
+    std::stringstream ss;
+    ss << "Unknown type of matrix '" << node.Tag() << "'.";
+    throw runtime_error(ss.str());
+  }
+}
+
+//==============================================================================
+template <typename MatrixType, bool IsVectorAtCompileTime>
+struct encode_impl
+{
+  // Nothing defined. This class should be always specialized.
+
+  // The reason this exists is so we can use template specialization to switch
+  // between serializing vectors and serializing matrices based on the
+  // IsVectorAtCompileTime flag. This is a "nice to have" that lets us generate
+  // !Vector [1, 2, 3] instead of the syntactic travesty like
+  // !Matrix [[1], [2] ,[3]] when serializing an Eigen vector.
+};
+
+//==============================================================================
+// Specialization for vector type
+template <typename MatrixType>
+struct encode_impl<MatrixType, true>
+{
+  static YAML::Node encode(const MatrixType& matrix)
+  {
+    using Index = typename MatrixType::Index;
+
+    YAML::Node node;
+    node.SetTag("Vector");
+
+    for (Index i = 0; i < matrix.size(); ++i)
+      node.push_back(YAML::Node(matrix[i]));
+
+    return node;
+  }
+};
+
+//==============================================================================
+// Specialization for matrix type
+template <typename MatrixType>
+struct encode_impl<MatrixType, false>
+{
+  static YAML::Node encode(const MatrixType& matrix)
+  {
+    using Index = typename MatrixType::Index;
+
+    YAML::Node node;
+    node.SetTag("Matrix");
+
+    for (Index r = 0; r < matrix.rows(); ++r)
+    {
+      YAML::Node row(YAML::NodeType::Sequence);
+
+      for (Index c = 0; c < matrix.cols(); ++c)
+        row.push_back(matrix(r, c));
+
+      node.push_back(row);
+    }
+
+    return node;
+  }
+};
+
+} // namespace (anonymous)
```

Comment:
I don't think we should _ever_ put an anonymous `namespace` in a header file. A header file can be `#include`ed into more than one translation unit, so this will result in unnecessary duplication of the objects (in the best case) or difficult-to-diagnose bugs (in the worst case, e.g. possibly if the namespace contains `static` member or global variables).

I see the `detail` namespace as the logical equivalent of an anonymous namespace in the header file, i.e. it is an implementation detail of other functions in that file.